File: api/utils/compression.py
"""Helpers for decoding compressed request payloads."""

import gzip
from typing import Optional

try:
    import brotli
except ModuleNotFoundError:  # pragma: no cover - optional dependency
    brotli = None


class UnsupportedEncodingError(ValueError):
    """Raised when the content encoding is not supported."""


def decompress_payload(content_encoding: Optional[str], payload: bytes) -> bytes:
    """Return uncompressed request bytes based on the Content-Encoding header."""

    if not payload:
        return payload

    if not content_encoding:
        return payload

    normalized = content_encoding.lower()
    if normalized == "gzip":
        return gzip.decompress(payload)
    if normalized in {"br", "brotli"}:
        if brotli is None:
            raise UnsupportedEncodingError("Brotli support not installed.")
        return brotli.decompress(payload)

    if normalized not in {"identity", "none"}:
        raise UnsupportedEncodingError(f"Unsupported content encoding: {content_encoding}")
    return payload

File: backend/src/utils/compression.ts
import { gunzipSync } from 'zlib';

export class UnsupportedEncodingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'UnsupportedEncodingError';
  }
}

export function decompressPayload(contentEncoding: string | undefined, payload: Buffer): Buffer {
  if (!payload || payload.length === 0) {
    return payload;
  }

  if (!contentEncoding) {
    return payload;
  }

  const normalized = contentEncoding.toLowerCase();

  if (normalized === 'gzip') {
    try {
      return gunzipSync(payload);
    } catch (error) {
      throw new Error('Invalid gzip payload');
    }
  }

  if (normalized === 'br' || normalized === 'brotli') {
    throw new UnsupportedEncodingError('Brotli support not implemented');
  }

  if (normalized !== 'identity' && normalized !== 'none') {
    throw new UnsupportedEncodingError(`Unsupported content encoding: ${contentEncoding}`);
  }

  return payload;
}
